File: lang/c/src/vpkppc/format/VPK_VTMB.cpp
#include <vpkppc/format/VPK_VTMB.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const char* VPKPP_VPK_VTMB_EXTENSION = VPK_VTMB_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_vpk_vtmb_create(const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = VPK_VTMB::create(path);
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_vpk_vtmb_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = VPK_VTMB::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

File: lang/csharp/src/vpkpp/Format/VPK_VTMB.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class VPK_VTMB : PackFile
{
	protected VPK_VTMB(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public static VPK_VTMB? Create(string path)
	{
		var handle = DLL.vpkpp_vpk_vtmb_create(path);
		return handle == nint.Zero ? null : new VPK_VTMB(handle);
	}

	public new static VPK_VTMB? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_vpk_vtmb_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new VPK_VTMB(handle);
	}
}
